src/sql-parser/src/parser.rs

```diff
@@ -1648,13 +1643,64 @@ impl<'a> Parser<'a> {
         self.expect_keyword(FROM)?;
         let connector = self.parse_connector()?;
         let with_options = self.parse_opt_with_sql_options()?;
-        let format = if self.parse_keyword(FORMAT) {
-            Some(self.parse_format()?)
+        // legacy upsert format syntax allows setting the key format after the keyword UPSERT, so we
+        // may mutate this variable in the next block
+        let mut format = if matches!(connector, Connector::Kafka { .. }) {
+            // pure.rs asserts that KeyValue formats only make sense for Kafka
+            match self.parse_one_of_keywords(&[KEY, FORMAT]) {
+                Some(KEY) => {
+                    self.expect_keyword(FORMAT)?;
+                    let key = self.parse_format()?;
+                    self.expect_keywords(&[VALUE, FORMAT])?;
+                    let value = self.parse_format()?;
+                    CreateSourceFormat::KeyValue { key, value }
+                }
+                Some(FORMAT) => CreateSourceFormat::Bare(self.parse_format()?),
+                Some(_) => unreachable!("parse_one_of_keywords returns None for this"),
+                None => CreateSourceFormat::None,
+            }
+        } else if self.parse_keyword(FORMAT) {
+            CreateSourceFormat::Bare(self.parse_format()?)
         } else {
-            None
+            CreateSourceFormat::None
         };
         let envelope = if self.parse_keyword(ENVELOPE) {
-            self.parse_envelope()?
+            let envelope = self.parse_envelope()?;
+            if matches!(envelope, Envelope::Upsert) {
+                // TODO: remove support for explicit UPSERT FORMAT after a period of deprecation
+                if self.parse_keyword(FORMAT) {
+                    warn!("UPSERT FORMAT has been deprecated, use the new KEY FORMAT syntax");
+                    if let CreateSourceFormat::Bare(value) = format {
+                        let key = self.parse_format()?;
+                        format = CreateSourceFormat::KeyValue { key, value };
+                    } else {
+                        self.error(
+                            self.index,
+                            "Key/Value format specified multiple times".into(),
```

Comment:
Let's say something like `ENVELOPE UPSERT FORMAT conflicts with earlier KEY FORMAT specification` to make it more clear which clauses are in conflict!